### PR TITLE
Fix for issues in Kevin's Fusion benchmark

### PR DESCRIPTION
**Pull request by jjsjann123**
1. fix segmentation fusion that drops group with outputs; (stolen from Christian)
2. patch shape_inference to drop obsolete rank inference;
3. disable permutation for reduction_to_size, which causes std::bad_alloc.

### CODE REVIEW

**Comment by jjsjann123** [comment]
dev_branch has failures, so it's tricky to test the fixes.
I'm investigating that. Kicking off the PR so we can at least unblock people who ran into these issues.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.h [hunk 1]
```diff
@@ -52,7 +52,7 @@ class TORCH_CUDA_CU_API SegmentedGroup {
 
   //! Checks if this group is used any where in the segmented fusion
   bool isConnected() const {
-    return !producer_edges.empty() || !consumer_edges.empty();
+    return !producer_edges.empty() || !consumer_edges.empty() || !output_vals.empty();
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.h]: @shmsong ^^^ this was dropping fusion group for me earlier. Stolen from Christian and seems to resolve the issue.

**Comment by jjsjann123** [comment]
Permutation related code in this PR is hacky. We need to overhaul/redesign that since now codegen supports complex reduction/broadcasting scheme in computation.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.h (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.h]: LGTM. Thanks for fixing this.

**Review by csarofeen — APPROVED** [review]

**Pull request merged**